File: vibeshield-python/examples/authorization_usage.py
"""
VibeShield Authorization & Access Control Protector — Usage Example

Demonstrates role verification, ownership checks, and static code scanning.

Run with: python examples/authorization_usage.py
"""

import os
import sys
import asyncio

# Allow importing from src/
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from src import (
    check_permission,
    validate_resource_ownership,
    detect_missing_auth_middleware,
    VibeShieldAuthorizationError,
    require_role,
    require_permission,
    require_ownership,
    require_auth,
)

print("=============================================================")
print("[SECURE] VIBESHIELD AUTHORIZATION PROTECTOR DEMO")
print("=============================================================\n")

# --- 1. Basic Permission and Ownership Checks ---
print("--- 1. Using check_permission() and validate_resource_ownership() ---")

users = {
    "admin_user": {"role": "admin", "id": 1},
    "regular_user": {"role": "user", "id": 100, "permissions": ["users:read"]},
    "guest_user": {"role": "guest", "id": 999}
}

# Check permissions
print("Permission Matrix Checks:")
for name, user in users.items():
    can_write_post = check_permission(user, "posts", "write")
    can_delete_user = check_permission(user, "users", "delete")
    print(f"  [{name}] Role: {user['role']}")
    print(f"    - Can write post? {can_write_post}")
    print(f"    - Can delete user? {can_delete_user}")

# Check ownership (IDOR protection)
print("\nOwnership Checks:")
resource_owner_id = 100  # owned by regular_user
for name, user in users.items():
    is_owner = validate_resource_ownership(
        user_id=user["id"],
        resource_id=5000,
        resource_owner_id=resource_owner_id,
        user_role=user["role"]
    )
    print(f"  [{name}] user_id: {user['id']} (role: {user['role']}) owns resource owned by 100? {is_owner}")


# --- 2. Middleware / Decorators Demonstration ---
print("\n--- 2. Route Protection Decorators ---")

# Define protected endpoints
@require_role("admin")
def delete_system_logs(user):
    return "SUCCESS: System logs deleted."

@require_permission("posts", "write")
def create_blog_post(user):
    return "SUCCESS: Blog post created."

@require_ownership("post_id")
def edit_blog_post(post_id: int, user: dict, owner_id: int):
    return f"SUCCESS: Post {post_id} edited."

@require_auth(roles=["user"], permissions=["posts:write"])
def publish_post(user):
    return "SUCCESS: Post published."

# Invoking sync decorators
print("\nExecuting sync decorated functions:")

# Test role decorator
try:
    print("  Calling delete_system_logs as regular_user...")
    delete_system_logs(user=users["regular_user"])
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")

try:
    print("  Calling delete_system_logs as admin_user...")
    res = delete_system_logs(user=users["admin_user"])
    print(f"  [ALLOWED] {res}")
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")

# Test permission decorator
try:
    print("  Calling create_blog_post as guest_user...")
    create_blog_post(user=users["guest_user"])
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")

try:
    print("  Calling create_blog_post as regular_user...")
    res = create_blog_post(user=users["regular_user"])
    print(f"  [ALLOWED] {res}")
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")

# Test ownership decorator
try:
    print("  Editing post owned by 100 as guest_user (id 999)...")
    edit_blog_post(post_id=456, user=users["guest_user"], owner_id=100)
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")

try:
    print("  Editing post owned by 100 as regular_user (id 100)...")
    res = edit_blog_post(post_id=456, user=users["regular_user"], owner_id=100)
    print(f"  [ALLOWED] {res}")
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")

# Test combined decorator
try:
    print("  Publishing post as regular_user...")
    res = publish_post(user={"role": "user", "permissions": ["posts:write"]})
    print(f"  [ALLOWED] {res}")
except VibeShieldAuthorizationError as e:
    print(f"  [BLOCKED] {e}")


# --- 3. Static Scanner Demonstration ---
print("\n--- 3. Static Authorization Scanner ---")

unsafe_code = """
@app.route("/admin/delete-database")
def delete_db():
    # Dangerous admin endpoint without auth guard!
    db.drop_all()

@app.get("/api/users/{user_id}")
def get_user_profile(user_id: int):
    # IDOR Vulnerable! Missing ownership check
    return db.query_user(user_id)
"""

safe_code = """
@require_role("admin")
@app.route("/admin/delete-database")
def delete_db():
    db.drop_all()

@require_ownership("user_id")
@app.get("/api/users/{user_id}")
def get_user_profile(user_id: int):
    return db.query_user(user_id)
"""

print("Scanning Unsafe Code Block:")
scan_result = detect_missing_auth_middleware(unsafe_code)
print(f"  Safe: {scan_result['safe']}")
for finding in scan_result["findings"]:
    print(f"  - [{finding['severity'].upper()}] Line {finding['line']}: {finding['pattern']}")
    print(f"    Code: {finding['original_code']}")
    print(f"    Suggestion: {finding['suggestion']}")

print("\nScanning Safe Code Block:")
scan_result_safe = detect_missing_auth_middleware(safe_code)
print(f"  Safe: {scan_result_safe['safe']}")
print(f"  Findings Count: {len(scan_result_safe['findings'])}")

File: vibeshield-ts/examples/authorization-usage.ts
/**
 * VibeShield Authorization & Access Control Protector — Usage Example
 *
 * This file demonstrates how to use the checkPermission, validateResourceOwnership,
 * requireAuth middleware, and detectMissingAuthMiddleware static scanner.
 *
 * Run with: npx tsx examples/authorization-usage.ts
 */

import {
  checkPermission,
  validateResourceOwnership,
  detectMissingAuthMiddleware,
  requireAuth,
  UserContext
} from '../src/index.js';

console.log('=============================================================');
console.log('🛡️ VIBESHIELD AUTHORIZATION & ACCESS CONTROL DEMO');
console.log('=============================================================\n');

// 1. Context definitions
const adminUser: UserContext = {
  id: 1,
  role: 'admin',
  permissions: []
};

const standardUser: UserContext = {
  id: 2,
  role: 'user',
  permissions: ['posts:write', 'settings:read']
};

const guestUser: UserContext = {
  id: 3,
  role: 'guest',
  permissions: []
};

console.log('--- 1. Role & Permission Matrix Checks ---');
console.log(`Admin access to "users:delete": ${checkPermission(adminUser, 'users', 'delete') ? '✅ Allowed' : '❌ Denied'}`);
console.log(`Standard User access to "posts:write": ${checkPermission(standardUser, 'posts', 'write') ? '✅ Allowed' : '❌ Denied'}`);
console.log(`Standard User access to "posts:delete": ${checkPermission(standardUser, 'posts', 'delete') ? '✅ Allowed' : '❌ Denied'}`);
console.log(`Guest access to "posts:read": ${checkPermission(guestUser, 'posts', 'read') ? '✅ Allowed' : '❌ Denied'}`);
console.log(`Guest access to "settings:read": ${checkPermission(guestUser, 'settings', 'read') ? '✅ Allowed' : '❌ Denied'}`);
console.log();

console.log('--- 2. Direct/Wildcard Permission Overrides ---');
const wildcardUser: UserContext = {
  id: 4,
  role: 'guest',
  permissions: ['settings:*']
};
console.log(`Wildcard user (guest + "settings:*") access to "settings:admin": ${checkPermission(wildcardUser, 'settings', 'admin') ? '✅ Allowed' : '❌ Denied'}`);
console.log();

console.log('--- 3. IDOR / Resource Ownership Validation ---');
const ownerId = 42;
const nonOwnerId = 99;
console.log(`Owner (User 42) accessing resource 100 owned by 42: ${validateResourceOwnership(ownerId, 100, ownerId) ? '✅ Allowed' : '❌ Denied'}`);
console.log(`Non-Owner (User 99) accessing resource 100 owned by 42: ${validateResourceOwnership(nonOwnerId, 100, ownerId) ? '✅ Allowed' : '❌ Denied'}`);
console.log(`Admin bypassing resource ownership: ${validateResourceOwnership(nonOwnerId, 100, ownerId, 'admin') ? '✅ Allowed (Bypassed)' : '❌ Denied'}`);
console.log();

console.log('--- 4. Code Scanning for Missing Auth Middleware ---');
const unsafeExpressCode = `
const express = require('express');
const router = express.Router();

// Critical: Admin route missing auth checks
router.get('/admin/dashboard', (req, res) => {
  res.send('Dashboard');
});

// Warning: User endpoint missing ownership checks
router.get('/api/users/:id', (req, res) => {
  res.send('User info');
});
`;

const scanResult = detectMissingAuthMiddleware(unsafeExpressCode);
console.log(`Scan Safe: ${scanResult.safe ? '✅ Yes' : '❌ No'}`);
console.log(`Findings count: ${scanResult.findings.length}`);
scanResult.findings.forEach(f => {
  console.log(`- [Line ${f.line}] [${f.severity.toUpperCase()}] ${f.pattern}`);
  console.log(`  Code: "${f.originalCode}"`);
  console.log(`  Suggestion: ${f.suggestion}`);
});
console.log();

console.log('--- 5. Middleware Simulation (Express & Next.js) ---');
// Simulating an Express middleware run - IDOR Blocked (User 2 tries to access resource owned by 42)
const reqBlocked = {
  user: standardUser,
  params: { id: '100' },
  body: { ownerId: '42' }
};

try {
  const expressMiddleware = requireAuth({ roles: ['user'], resourceIdParam: 'id' });
  expressMiddleware(reqBlocked, {}, () => {
    console.log('✅ Express Middleware passed unexpectedly.');
  });
} catch (err: any) {
  console.log(`🛡️ Express Middleware blocked unauthorized access: ${err.message}`);
}

// Simulating an Express middleware run - Allowed (User 2 tries to access resource owned by 2)
const reqAllowed = {
  user: standardUser,
  params: { id: '100' },
  body: { ownerId: '2' }
};

try {
  const expressMiddleware = requireAuth({ roles: ['user'], resourceIdParam: 'id' });
  expressMiddleware(reqAllowed, {}, () => {
    console.log('✅ Express Middleware passed for User 2 accessing own resource.');
  });
} catch (err: any) {
  console.log(`❌ Express Middleware failed: ${err.message}`);
}

